From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: suvojeet-sengupta <suvojitsengupta21@gmail.com>
Date: Tue, 6 Oct 2026 23:13:02 +0000
Subject: [PATCH] base: Add three finger swipe to screenshot [1/2]

Swiping down with three fingers takes a screenshot when
Settings.System.THREE_FINGER_GESTURE is on. A pointer event listener on
the default display watches for three pointers that each move 100dp
down within half a second.

Change-Id: I9e76c6e7211e217f8737931027aa9f45399b8880
---
 core/java/android/provider/Settings.java      |  6 ++
 .../validators/SystemSettingsValidators.java  |  1 +
 .../server/policy/PhoneWindowManager.java     | 24 +++++++
 .../policy/SwipeToScreenshotListener.java     | 70 +++++++++++++++++++
 4 files changed, 101 insertions(+)
 create mode 100644 services/core/java/com/android/server/policy/SwipeToScreenshotListener.java

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index 68c48fe4..5c926f20 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -7035,6 +7035,12 @@ public final class Settings {
          */
         public static final String GESTURE_DOUBLE_TAP = "gesture_double_tap";
 
+        /**
+         * Whether swiping down with three fingers takes a screenshot.
+         * @hide
+         */
+        public static final String THREE_FINGER_GESTURE = "three_finger_gesture";
+
         /**
          * Controls whether single tap to wake is enabled.
          * @hide
diff --git a/packages/SettingsProvider/src/android/provider/settings/validators/SystemSettingsValidators.java b/packages/SettingsProvider/src/android/provider/settings/validators/SystemSettingsValidators.java
index 90903d14..010dae18 100644
--- a/packages/SettingsProvider/src/android/provider/settings/validators/SystemSettingsValidators.java
+++ b/packages/SettingsProvider/src/android/provider/settings/validators/SystemSettingsValidators.java
@@ -288,6 +288,7 @@ public class SystemSettingsValidators {
                 new PackageNameListValidator(","));
         VALIDATORS.put(System.GESTURES_ENABLED, BOOLEAN_VALIDATOR);
         VALIDATORS.put(System.GESTURE_DOUBLE_TAP, NON_NEGATIVE_INTEGER_VALIDATOR);
+        VALIDATORS.put(System.THREE_FINGER_GESTURE, BOOLEAN_VALIDATOR);
         VALIDATORS.put(System.GESTURE_SINGLE_TAP, NON_NEGATIVE_INTEGER_VALIDATOR);
         VALIDATORS.put(System.GESTURE_DRAW_V, NON_NEGATIVE_INTEGER_VALIDATOR);
         VALIDATORS.put(System.GESTURE_DRAW_INVERSE_V, NON_NEGATIVE_INTEGER_VALIDATOR);
diff --git a/services/core/java/com/android/server/policy/PhoneWindowManager.java b/services/core/java/com/android/server/policy/PhoneWindowManager.java
index b7ae2aba..9be8eccf 100644
--- a/services/core/java/com/android/server/policy/PhoneWindowManager.java
+++ b/services/core/java/com/android/server/policy/PhoneWindowManager.java
@@ -239,6 +239,8 @@ import com.android.internal.policy.KeyInterceptionInfo;
 import com.android.internal.policy.PhoneWindow;
 import com.android.internal.statusbar.IStatusBarService;
 import com.android.internal.util.NamedLock;
+import com.android.internal.util.ScreenshotHelper;
+import com.android.internal.util.ScreenshotRequest;
 import com.android.internal.widget.LockPatternUtils;
 import com.android.server.AccessibilityManagerInternal;
 import com.android.server.DockObserverInternal;
@@ -514,6 +516,9 @@ public class PhoneWindowManager implements WindowManagerPolicy {
     private GlobalActions mGlobalActions;
     private Handler mHandler;
 
+    private volatile boolean mThreeFingerGestureEnabled;
+    private ScreenshotHelper mScreenshotHelper;
+
     // FIXME This state is shared between the input reader and handler thread.
     // Technically it's broken and buggy but it has been like this for many years
     // and we have not yet seen any problems.  Someday we'll rewrite this logic
@@ -913,6 +918,9 @@ public class PhoneWindowManager implements WindowManagerPolicy {
             resolver.registerContentObserver(Settings.Secure.getUriFor(
                     Settings.Secure.INCALL_BACK_BUTTON_BEHAVIOR), false, this,
                     UserHandle.USER_ALL);
+            resolver.registerContentObserver(Settings.System.getUriFor(
+                    Settings.System.THREE_FINGER_GESTURE), false, this,
+                    UserHandle.USER_ALL);
             resolver.registerContentObserver(Settings.Secure.getUriFor(
                     Settings.Secure.WAKE_GESTURE_ENABLED), false, this,
                     UserHandle.USER_ALL);
@@ -3093,6 +3101,9 @@ public class PhoneWindowManager implements WindowManagerPolicy {
             mSystemNavigationKeysEnabled = Settings.Secure.getIntForUser(resolver,
                     Settings.Secure.SYSTEM_NAVIGATION_KEYS_ENABLED,
                     0, UserHandle.USER_CURRENT) == 1;
+            mThreeFingerGestureEnabled = Settings.System.getIntForUser(resolver,
+                    Settings.System.THREE_FINGER_GESTURE,
+                    0, UserHandle.USER_CURRENT) == 1;
             mRingerToggleChord = Settings.Secure.getIntForUser(resolver,
                     Settings.Secure.VOLUME_HUSH_GESTURE, VOLUME_HUSH_OFF,
                     UserHandle.USER_CURRENT);
@@ -6238,6 +6249,15 @@ public class PhoneWindowManager implements WindowManagerPolicy {
         bindKeyguard();
     }
 
+    private void takeThreeFingerScreenshot() {
+        if (!mThreeFingerGestureEnabled) {
+            return;
+        }
+        mScreenshotHelper.takeScreenshot(new ScreenshotRequest.Builder(
+                WindowManager.TAKE_SCREENSHOT_FULLSCREEN,
+                WindowManager.ScreenshotSource.SCREENSHOT_OTHER).build(), mHandler, null);
+    }
+
     /** {@inheritDoc} */
     @Override
     public void systemReady() {
@@ -6249,6 +6269,10 @@ public class PhoneWindowManager implements WindowManagerPolicy {
         mPocketManager.addCallback(mPocketCallback);
         mPocketLock = new PocketLock(mContext);
 
+        mScreenshotHelper = new ScreenshotHelper(mContext);
+        mWindowManagerFuncs.registerPointerEventListener(new SwipeToScreenshotListener(mContext,
+                () -> mHandler.post(this::takeThreeFingerScreenshot)), DEFAULT_DISPLAY);
+
         mVrManagerInternal = LocalServices.getService(VrManagerInternal.class);
         if (mVrManagerInternal != null) {
             mVrManagerInternal.addPersistentVrModeStateListener(mPersistentVrModeListener);
diff --git a/services/core/java/com/android/server/policy/SwipeToScreenshotListener.java b/services/core/java/com/android/server/policy/SwipeToScreenshotListener.java
new file mode 100644
index 00000000..d915f3d3
--- /dev/null
+++ b/services/core/java/com/android/server/policy/SwipeToScreenshotListener.java
@@ -0,0 +1,70 @@
+/*
+ * SPDX-FileCopyrightText: 2026 Paranoid Android
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.android.server.policy;
+
+import android.content.Context;
+import android.view.MotionEvent;
+import android.view.WindowManagerPolicyConstants.PointerEventListener;
+
+/** Calls back when three fingers swipe down together. */
+class SwipeToScreenshotListener implements PointerEventListener {
+
+    private static final int POINTER_COUNT = 3;
+    private static final long MAX_SWIPE_TIME_MS = 500;
+    private static final float SWIPE_DISTANCE_DP = 100;
+
+    private final Runnable mCallback;
+    private final float mSwipeDistance;
+    private final int[] mPointerIds = new int[POINTER_COUNT];
+    private final float[] mStartY = new float[POINTER_COUNT];
+    private boolean mTracking;
+    private long mStartTime;
+
+    SwipeToScreenshotListener(Context context, Runnable callback) {
+        mCallback = callback;
+        mSwipeDistance = SWIPE_DISTANCE_DP * context.getResources().getDisplayMetrics().density;
+    }
+
+    @Override
+    public void onPointerEvent(MotionEvent event) {
+        switch (event.getActionMasked()) {
+            case MotionEvent.ACTION_POINTER_DOWN:
+                mTracking = event.getPointerCount() == POINTER_COUNT;
+                if (mTracking) {
+                    for (int i = 0; i < POINTER_COUNT; i++) {
+                        mPointerIds[i] = event.getPointerId(i);
+                        mStartY[i] = event.getY(i);
+                    }
+                    mStartTime = event.getEventTime();
+                }
+                break;
+            case MotionEvent.ACTION_MOVE:
+                if (!mTracking) {
+                    break;
+                }
+                if (event.getPointerCount() != POINTER_COUNT
+                        || event.getEventTime() - mStartTime > MAX_SWIPE_TIME_MS) {
+                    mTracking = false;
+                    break;
+                }
+                for (int i = 0; i < POINTER_COUNT; i++) {
+                    final int index = event.findPointerIndex(mPointerIds[i]);
+                    if (index < 0 || event.getY(index) - mStartY[i] < mSwipeDistance) {
+                        return;
+                    }
+                }
+                mTracking = false;
+                mCallback.run();
+                break;
+            case MotionEvent.ACTION_DOWN:
+            case MotionEvent.ACTION_POINTER_UP:
+            case MotionEvent.ACTION_UP:
+            case MotionEvent.ACTION_CANCEL:
+                mTracking = false;
+                break;
+        }
+    }
+}
-- 
2.55.0

